src/effects/backends/builtin/compressoreffect.cpp

```diff
@@ -0,0 +1,215 @@
+    attack->setRange(0, 10, 250);
+
+    EffectManifestParameterPointer release = pManifest->addParameter();
+    release->setId("release");
+    release->setName(QObject::tr("Release (ms)"));
+    release->setShortName(QObject::tr("Release"));
+    release->setDescription(QObject::tr(
+        "The Release knob sets the time that determines how fast the compressor will recover from the gain reduction once the signal falls under the threshold. "
+        "Depending on the input signal, short release times may introduce a 'pumping' effect and/or distortion"));
+    release->setValueScaler(EffectManifestParameter::ValueScaler::Integral);
+    release->setUnitsHint(EffectManifestParameter::UnitsHint::Millisecond);
+    release->setRange(0, 150, 1500);
+
+    EffectManifestParameterPointer gain = pManifest->addParameter();
+    gain->setId("gain");
+    gain->setName(QObject::tr("Output gain"));
+    gain->setShortName(QObject::tr("Gain"));
+    gain->setDescription(QObject::tr("The Output gain knob adjusts the level of the output signal after the compression was applied"));
+    gain->setValueScaler(EffectManifestParameter::ValueScaler::Linear);
+    gain->setUnitsHint(EffectManifestParameter::UnitsHint::Decibel);
+    gain->setRange(-25, 0, 25);
+
+    return pManifest;
+}
+
+CompressorGroupState::CompressorGroupState(
+        const mixxx::EngineParameters& engineParameters)
+        : EffectState(engineParameters),
+          previousMakeUpGain(0),
+          previousStateDB(0) {
+}
+
+void CompressorEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pThreshold = parameters.value("threshold");
+    m_pRatio = parameters.value("ratio");
+    m_pKnee = parameters.value("knee");
+    m_pAttack = parameters.value("attack");
+    m_pRelease = parameters.value("release");
+    m_pGain = parameters.value("gain");
+    m_pAutoMakeUp = parameters.value("automakeup");
+    m_pClipping = parameters.value("clipping");
+}
+
+void CompressorEffect::processChannel(
+        CompressorGroupState* pState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(enableState);
+
+    SINT numSamples = engineParameters.samplesPerBuffer();
+
+    // Compression
+    applyCompression(pState, engineParameters, pInput, pOutput);
+
+    // Auto make up
+    if (m_pAutoMakeUp->toInt() == AutoMakeUpOn) { 
+        applyAutoMakeUp(pState, pOutput, numSamples);
+    }
+
+    // Output gain
+    CSAMPLE gainParamDB = static_cast<CSAMPLE>(m_pGain->value());
+    SampleUtil::applyGain(pOutput, db2ratio(gainParamDB), numSamples);
+
+    // Clipping
+    if (m_pClipping->toInt() == ClippingOn) {
+        SampleUtil::copyClampBuffer(pOutput, pOutput, numSamples);
+    }
```

Comment:
`SampleUtil::copyClampBuffer` can't be run in place (due to `-Wrestrict`). Feel free to create an inplace-clamp function to `SampleUtil`.

Reply:
Done